src/clib/pioc_support.c

```diff
@@ -965,10 +969,100 @@ int PIOc_write_nc_decomp(const char *filename, int iosysid, int ioid, MPI_Comm c
+ * NULL.
+ * @param fortran_order pointer that gets set to 1 if fortran array
+ * ordering is used, or to zero if C array ordering is used.
+ * @returns 0 for success, error code otherwise.
+ */
+int PIOc_read_nc_decomp(const char *filename, int iosysid, int *ioidp, MPI_Comm comm,
+                        char *title, char *history, int *fortran_order)
+{
+    iosystem_desc_t *ios; /* Pointer to the IO system info. */
+    int ndims;            /* The number of data dims (except unlim). */
+    int max_maplen;       /* The max maplen of any task. */
+    int *global_dimlen;   /* An array with sizes of global dimensions. */
+    int *task_maplen;     /* A map of one tasks mapping to global data. */
+    int *full_map;        /* A map with the task maps of every task. */
+    int num_tasks_decomp; /* The number of tasks for this decomp. */
+    int size;             /* Size of comm. */
+    int my_rank;          /* Task rank in comm. */
+    char source_in[PIO_MAX_NAME + 1];  /* Text metadata in decomp file. */
+    char version_in[PIO_MAX_NAME + 1]; /* Text metadata in decomp file. */
+    int basetype = PIO_INT;
+    int mpierr;
+    int ret;
+
+    /* Get the IO system info. */
+    if (!(ios = pio_get_iosystem_from_id(iosysid)))
+        return pio_err(NULL, NULL, PIO_EBADID, __FILE__, __LINE__);
+
+    /* Check inputs. */
+    if (!filename || !ioidp)
+        return pio_err(ios, NULL, PIO_EINVAL, __FILE__, __LINE__);
+
+    LOG((1, "PIOc_read_nc_decomp filename = %s iosysid = %d", filename, iosysid));
+
+    /* Get the communicator size and task rank. */
+    if ((mpierr = MPI_Comm_size(comm, &size)))
+        return check_mpi2(ios, NULL, mpierr, __FILE__, __LINE__);
+    if ((mpierr = MPI_Comm_rank(comm, &my_rank)))
+        return check_mpi2(ios, NULL, mpierr, __FILE__, __LINE__);
+    LOG((2, "size = %d my_rank = %d", size, my_rank));
+
+    /* Read the file. This allocates three arrays that we have to
+     * free. */
+    if ((ret = pioc_read_nc_decomp_int(iosysid, filename, &ndims, &global_dimlen, &num_tasks_decomp,
+                                       &task_maplen, &max_maplen, &full_map, title, history,
+                                       source_in, version_in, fortran_order)))
+        return ret;
+    LOG((2, "ndims = %d num_tasks_decomp = %d max_maplen = %d", ndims, num_tasks_decomp,
+         max_maplen));
+
+    /* If the size does not match the number of tasks in the decomp,
+     * that's an error. */
+    if (size != num_tasks_decomp)
+        ret = PIO_EINVAL;
+
+    /* Now initialize the iodesc on each task for this decomposition. */
+    if (!ret)
+    {
+        PIO_Offset compmap[task_maplen[my_rank]];
+
+        /* Copy array into PIO_Offset array. */
+        for (int e = 0; e < task_maplen[my_rank]; e++)
+            compmap[e] = full_map[my_rank * max_maplen + e];
+
+        /* Initialize the decomposition. */
+        ret = PIOc_InitDecomp(iosysid, basetype, ndims, global_dimlen, task_maplen[my_rank],
```

Comment:
It's not apparent to me that we would want to do this here rather than let the user do it?